packages/teuchos/parameterlist/src/Teuchos_ParameterList.hpp

```diff
@@ -260,9 +260,9 @@ class TEUCHOSPARAMETERLIST_LIB_DLL_EXPORT ParameterList {
     \note This fallback is necessary to support legacy use cases that specify the type
           of the parameter at the call site, and the type is hence not deduced.  
   */
-  template<typename T, typename S, typename = std::enable_if_t< ! std::is_same_v<T, S> && std::is_same_v<T, std::decay_t<S>>>>
+  template<typename T, typename S, typename = std::enable_if_t<std::is_convertible_v<S, std::remove_const_t<T>>>>
   ParameterList& set (std::string const& name,
-                      S&& value,
+                      const S& value,
```

Comment:
@maartenarnst, the original purpose of PR #13409 was:

> The intent of this change is to allow move semantics along the chain of calls that is made when setting a parameter.

By passing the value by `const S&`, does not not completely break the original intent of these changes which was to improve the performance of setting large objects in a PL?

Reply:
I don't think so. This change affects only the fallback. The main version of the templated set method, which passes by universal ref, is still there. Hence, I think that the move semantics remain.

Reply:
I thing I tend to agree with @maartenarnst.

You still get the move semantics guarantee if you let the type be deduced at the call side (see the tests introduced in the previous PR, hope they still pass).

Here, @maartenarnst modifies the "set when type is specified at call site" to allow for user-defined conversions.

I think the tests should cover the following cases:
1. Set by `const&`, no conversion, type deduced.
2. Set by `const&`, no conversion, type specified at call site.
3. Set by `const&`, conversion, type deduced.
4. Set by `const&`, conversion, type specified.
5. Set by move semantics, type deduced.
6. Set by move semantics, type specified.

If this PR can prove these use cases are covered, I think it's impossible to break any downstream code.

Reply:
> I don't think so. This change affects only the fallback. The main version of the templated set method, which passes by universal ref, is still there. Hence, I think that the move semantics remain.

That is a good point.  And the new unit tests you added in PR #13409 does seem to show that.

It is only the optional overload:

```
  template<typename T, typename S, typename = std::enable_if_t<std::is_convertible_v<S, std::remove_const_t<T>>>>
  ParameterList& set (std::string const& name,
                      const S& value,
                      std::string const& docString = "",
                      RCP<const ParameterEntryValidator> const& validator = null);
```

that is always going to be doing a conversion of the input `value` object since it's type `S` does not match the stored type `T`.  If the argument type and the stored match, then the first function will be called that uses a universal reference (which supports move semantics):

```
  template<typename T>
  ParameterList& set (std::string const& name,
                      T&& value,
                      std::string const& docString = "",
                      RCP<const ParameterEntryValidator> const& validator = null);
```

I will go ahead and run the PR builds on this and see how this looks.

Thanks for being patient with me :-)